File: server/src/main/java/spring/serverspringboot/service/PartnerService.java
package spring.serverspringboot.service;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import spring.serverspringboot.dto.request.PartnerRequest;
import spring.serverspringboot.dto.response.PartnerResponse;
import spring.serverspringboot.entity.Partner;
import spring.serverspringboot.exception.AppException;
import spring.serverspringboot.exception.ErrorCode;
import spring.serverspringboot.mapper.PartnerMapper;
import spring.serverspringboot.repository.PartnerRepository;

import java.util.List;

@Service
@Slf4j
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class PartnerService {
    PartnerRepository partnerRepository;
    PartnerMapper partnerMapper;


// Customer

    public PartnerResponse createCustomer(PartnerRequest request) {
        Partner partner = partnerRepository.findPartnerByPartnerCode(request.getPartnerCode()).orElse(null);
        if (partner != null) {
            if (partner.isCustomer()) {
                throw new AppException(ErrorCode.PARTNER_EXISTED);
            } else {
                partner.setCustomer(true);
                partner.setActive(true);
                partnerRepository.save(partner);
            }
        } else {
            partner = partnerMapper.toPartner(request);
            partner.setCustomer(true);
            partner.setActive(true);
            partnerRepository.save(partner);
        }
        return partnerMapper.toPartnerResponse(partner);
    }
    public PartnerResponse getCustomer(Long partnerId){
        Partner partner = partnerRepository.findPartnerByIdAndIsCustomer(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
        return partnerMapper.toPartnerResponse(partner);
    }
    public List<Partner> getCustomers(){
        return partnerRepository.findAllByIsCustomerTrue();
    }

    public PartnerResponse updateCustomer(Long partnerId, PartnerRequest request) {
        Partner partner = partnerRepository.findPartnerByIdAndIsCustomer(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
        partnerMapper.updatePartner(partner, request);
        partner.setCustomer(true);
        partner = partnerRepository.save(partner);
        return partnerMapper.toPartnerResponse(partner);
    }

    public void deleteCustomer(Long partnerId) {
        Partner partner = partnerRepository.findPartnerByIdAndIsCustomer(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
        partner.setCustomer(false);

        if (!partner.isSupplier()) {
            partner.setActive(false);
        }

        partnerRepository.save(partner);
    }

// Supplier

    public PartnerResponse createSupplier(PartnerRequest request) {
        Partner partner = partnerRepository.findPartnerByPartnerCode(request.getPartnerCode()).orElse(null);
        if (partner != null) {
            if (partner.isSupplier()) {
                throw new AppException(ErrorCode.PARTNER_EXISTED);
            } else {
                partner.setSupplier(true);
                partner.setActive(true);
                partnerRepository.save(partner);
            }
        } else {
            partner = partnerMapper.toPartner(request);
            partner.setSupplier(true);
            partner.setActive(true);
            partnerRepository.save(partner);
        }
        return partnerMapper.toPartnerResponse(partner);
    }
    public PartnerResponse getSupplier(Long partnerId){
        Partner partner = partnerRepository.findPartnerByIdAndIsSupplier(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
        return partnerMapper.toPartnerResponse(partner);
    }
    public List<Partner> getSuppliers(){
        return partnerRepository.findAllByIsSupplierTrue();
    }

    public PartnerResponse updateSupplier(Long partnerId, PartnerRequest request) {
        Partner partner = partnerRepository.findPartnerByIdAndIsSupplier(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
        partnerMapper.updatePartner(partner, request);
        partner.setSupplier(true);
        partner = partnerRepository.save(partner);
        return partnerMapper.toPartnerResponse(partner);
    }

    public void deleteSupplier(Long partnerId) {
        Partner partner = partnerRepository.findPartnerByIdAndIsSupplier(partnerId, true)
                .orElseThrow(() -> new AppException(ErrorCode.PARTNER_NOT_EXISTED));
         partner.setSupplier(false);

        if (!partner.isCustomer()) {
            partner.setActive(false);
        }
        partnerRepository.save(partner);
    }
}

File: client/src/services/partner.service.ts
import { apiClient } from '../lib/api-client';

export interface Partner {
  id: number;
  partnerCode: string;
  name: string;
  isCustomer: boolean;
  isSupplier: boolean;
  address: string;
  taxCode: string;
  phone: string;
}

export const partnerService = {
  getCustomers: async (): Promise<Partner[]> => {
    const res = await apiClient.get('/partners/customers');
    return res.data.result;
  },

  createCustomer: async (partner: Partial<Partner>): Promise<Partner> => {
    const res = await apiClient.post('/partners/customers', partner);
    return res.data.result;
  },

  updateCustomer: async (id: number, partner: Partial<Partner>): Promise<Partner> => {
    const res = await apiClient.put(`/partners/customers/${id}`, partner);
    return res.data.result;
  },

  deleteCustomer: async (id: number): Promise<void> => {
    await apiClient.delete(`/partners/customers/${id}`);
  },

  getSuppliers: async (): Promise<Partner[]> => {
    const res = await apiClient.get('/partners/suppliers');
    return res.data.result;
  },

  createSupplier: async (partner: Partial<Partner>): Promise<Partner> => {
    const res = await apiClient.post('/partners/suppliers', partner);
    return res.data.result;
  },

  updateSupplier: async (id: number, partner: Partial<Partner>): Promise<Partner> => {
    const res = await apiClient.put(`/partners/suppliers/${id}`, partner);
    return res.data.result;
  },

  deleteSupplier: async (id: number): Promise<void> => {
    await apiClient.delete(`/partners/suppliers/${id}`);
  },
};
